File: ouro_sdk_js/src/microchain.ts
import { OuroClient } from './client';
import { Transaction, TransactionBuilder } from './transaction';
import { InvalidConfigError, TransactionFailedError } from './errors';
import type {
  MicrochainConfig,
  MicrochainState,
  BlockHeader,
  TransactionData,
  TxHistoryResponse,
  BlocksResponse,
  ConsensusType,
  AnchorFrequency,
} from './types';
import axios from 'axios';

/**
 * Microchain interface for building dApps
 */
export class Microchain {
  /** Microchain ID */
  public id: string;

  /** Client for node communication */
  private client: OuroClient;

  /** Current nonce for transactions (auto-incremented) */
  private nonce: number;

  /** Base URL for direct API calls */
  private baseUrl: string;

  private constructor(
    id: string,
    client: OuroClient,
    baseUrl: string,
    nonce: number = 0
  ) {
    this.id = id;
    this.client = client;
    this.baseUrl = baseUrl;
    this.nonce = nonce;
  }

  /**
   * Connect to an existing microchain
   */
  static async connect(
    microchainId: string,
    nodeUrl: string
  ): Promise<Microchain> {
    const client = new OuroClient(nodeUrl);

    // Verify microchain exists
    await client.getMicrochainState(microchainId);

    return new Microchain(microchainId, client, nodeUrl, 0);
  }

  /**
   * Create a new microchain
   */
  static async create(
    config: MicrochainConfig,
    nodeUrl: string
  ): Promise<Microchain> {
    const client = new OuroClient(nodeUrl);
    const id = await client.createMicrochain(config);

    return new Microchain(id, client, nodeUrl, 0);
  }

  /**
   * Get microchain state
   */
  async state(): Promise<MicrochainState> {
    return this.client.getMicrochainState(this.id);
  }

  /**
   * Get balance for an address on this microchain
   */
  async balance(address: string): Promise<number> {
    return this.client.getMicrochainBalance(this.id, address);
  }

  /**
   * Submit a transaction to this microchain
   */
  async submitTx(tx: Transaction): Promise<string> {
    try {
      const url = `${this.baseUrl}/microchain/${this.id}/tx`;
      const response = await axios.post(url, tx.toJSON());

      if (response.data.success) {
        this.nonce += 1;
        return response.data.tx_id;
      } else {
        throw new TransactionFailedError(
          response.data.message || 'Unknown error'
        );
      }
    } catch (error) {
      if (error instanceof TransactionFailedError) {
        throw error;
      }
      throw new TransactionFailedError(
        error instanceof Error ? error.message : 'Unknown error'
      );
    }
  }

  /**
   * Create a transaction builder for this microchain
   */
  tx(): TransactionBuilder {
    const builder = new TransactionBuilder();
    builder.setNonce(this.nonce);
    return builder;
  }

  /**
   * Transfer tokens on this microchain (simplified)
   */
  async transfer(from: string, to: string, amount: number): Promise<string> {
    const tx = new Transaction(from, to, amount);
    tx.nonce = this.nonce;

    return this.submitTx(tx);
  }

  /**
   * Anchor this microchain to subchain/mainchain
   */
  async anchor(): Promise<string> {
    return this.client.anchorMicrochain(this.id);
  }

  /**
   * Get transaction history for this microchain
   */
  async txHistory(from: number, to: number): Promise<TransactionData[]> {
    try {
      const url = `${this.baseUrl}/microchain/${this.id}/txs?from=${from}&to=${to}`;
      const response = await axios.get<TxHistoryResponse>(url);

      return response.data.transactions;
    } catch (error) {
      throw new Error(
        error instanceof Error ? error.message : 'Failed to fetch history'
      );
    }
  }

  /**
   * Get latest blocks from this microchain
   */
  async blocks(limit: number): Promise<BlockHeader[]> {
    try {
      const url = `${this.baseUrl}/microchain/${this.id}/blocks?limit=${limit}`;
      const response = await axios.get<BlocksResponse>(url);

      return response.data.blocks;
    } catch (error) {
      throw new Error(
        error instanceof Error ? error.message : 'Failed to fetch blocks'
      );
    }
  }
}

/**
 * Builder for creating microchains
 */
export class MicrochainBuilder {
  private config: MicrochainConfig;
  private nodeUrl?: string;

  constructor(name: string, owner: string) {
    this.config = {
      name,
      owner,
      consensus: {
        type: ConsensusType.SingleValidator,
      },
      anchorFrequency: { type: 'blocks', count: 100 },
      maxTxsPerBlock: 1000,
      blockTimeSecs: 5,
    };
  }

  /**
   * Set node URL
   */
  node(url: string): this {
    this.nodeUrl = url;
    return this;
  }

  /**
   * Set consensus type
   */
  consensus(
    type: ConsensusType,
    validatorCount?: number
  ): this {
    this.config.consensus = {
      type,
      validatorCount,
    };
    return this;
  }

  /**
   * Set anchor frequency
   */
  anchorFrequency(frequency: AnchorFrequency): this {
    this.config.anchorFrequency = frequency;
    return this;
  }

  /**
   * Set block time
   */
  blockTime(seconds: number): this {
    this.config.blockTimeSecs = seconds;
    return this;
  }

  /**
   * Build and create the microchain
   */
  async build(): Promise<Microchain> {
    if (!this.nodeUrl) {
      throw new InvalidConfigError('Node URL not specified');
    }

    return Microchain.create(this.config, this.nodeUrl);
  }
}

File: ouro_sdk_python/ouro_sdk/microchain.py
"""Microchain interface for building dApps"""

from typing import List, Dict, Any, Optional
import requests
from .client import OuroClient
from .transaction import Transaction, TransactionBuilder
from .types import (
    MicrochainConfig,
    MicrochainState,
    BlockHeader,
    TransactionData,
    ConsensusType,
    AnchorFrequency,
)
from .errors import InvalidConfigError, TransactionFailedError


class Microchain:
    """Microchain interface for building dApps"""

    def __init__(
        self, microchain_id: str, client: OuroClient, base_url: str, nonce: int = 0
    ):
        self.id = microchain_id
        self._client = client
        self._base_url = base_url
        self._nonce = nonce

    @classmethod
    def connect(cls, microchain_id: str, node_url: str) -> "Microchain":
        """Connect to an existing microchain"""
        client = OuroClient(node_url)

        # Verify microchain exists
        client.get_microchain_state(microchain_id)

        return cls(microchain_id, client, node_url, 0)

    @classmethod
    def create(cls, config: MicrochainConfig, node_url: str) -> "Microchain":
        """Create a new microchain"""
        client = OuroClient(node_url)
        microchain_id = client.create_microchain(config)

        return cls(microchain_id, client, node_url, 0)

    def state(self) -> MicrochainState:
        """Get microchain state"""
        return self._client.get_microchain_state(self.id)

    def balance(self, address: str) -> int:
        """Get balance for an address on this microchain"""
        return self._client.get_microchain_balance(self.id, address)

    def submit_tx(self, tx: Transaction) -> str:
        """Submit a transaction to this microchain"""
        try:
            url = f"{self._base_url}/microchain/{self.id}/tx"
            response = requests.post(url, json=tx.to_json().to_dict())
            response.raise_for_status()
            data = response.json()

            if data.get("success"):
                self._nonce += 1
                return data["tx_id"]
            else:
                raise TransactionFailedError(data.get("message", "Unknown error"))
        except TransactionFailedError:
            raise
        except Exception as e:
            raise TransactionFailedError(str(e))

    def tx(self) -> TransactionBuilder:
        """Create a transaction builder for this microchain"""
        builder = TransactionBuilder()
        builder.set_nonce(self._nonce)
        return builder

    def transfer(self, from_addr: str, to: str, amount: int) -> str:
        """Transfer tokens on this microchain (simplified)"""
        tx = Transaction(from_addr, to, amount)
        tx.nonce = self._nonce

        return self.submit_tx(tx)

    def anchor(self) -> str:
        """Anchor this microchain to subchain/mainchain"""
        return self._client.anchor_microchain(self.id)

    def tx_history(self, from_block: int, to_block: int) -> List[TransactionData]:
        """Get transaction history for this microchain"""
        try:
            url = f"{self._base_url}/microchain/{self.id}/txs?from={from_block}&to={to_block}"
            response = requests.get(url)
            response.raise_for_status()
            data = response.json()

            return [TransactionData.from_dict(tx) for tx in data["transactions"]]
        except Exception as e:
            raise Exception(f"Failed to fetch history: {str(e)}")

    def blocks(self, limit: int) -> List[BlockHeader]:
        """Get latest blocks from this microchain"""
        try:
            url = f"{self._base_url}/microchain/{self.id}/blocks?limit={limit}"
            response = requests.get(url)
            response.raise_for_status()
            data = response.json()

            return [BlockHeader.from_dict(block) for block in data["blocks"]]
        except Exception as e:
            raise Exception(f"Failed to fetch blocks: {str(e)}")


class MicrochainBuilder:
    """Builder for creating microchains"""

    def __init__(self, name: str, owner: str):
        self._config = MicrochainConfig(
            name=name,
            owner=owner,
            consensus={"type": ConsensusType.SINGLE_VALIDATOR.value},
            anchor_frequency=AnchorFrequency.every_n_blocks(100),
            max_txs_per_block=1000,
            block_time_secs=5,
        )
        self._node_url: Optional[str] = None

    def node(self, url: str) -> "MicrochainBuilder":
        """Set node URL"""
        self._node_url = url
        return self

    def consensus(
        self, consensus_type: ConsensusType, validator_count: Optional[int] = None
    ) -> "MicrochainBuilder":
        """Set consensus type"""
        self._config.consensus = {"type": consensus_type.value}
        if validator_count:
            self._config.consensus["validator_count"] = validator_count
        return self

    def anchor_frequency(self, frequency: AnchorFrequency) -> "MicrochainBuilder":
        """Set anchor frequency"""
        self._config.anchor_frequency = frequency
        return self

    def block_time(self, seconds: int) -> "MicrochainBuilder":
        """Set block time"""
        self._config.block_time_secs = seconds
        return self

    def build(self) -> Microchain:
        """Build and create the microchain"""
        if not self._node_url:
            raise InvalidConfigError("Node URL not specified")

        return Microchain.create(self._config, self._node_url)
